src/server/pfs/server/driver.go

```diff
@@ -3406,18 +3406,41 @@ func (d *driver) downloadTree(pachClient *client.APIClient, object *pfs.Object,
 	if err != nil {
 		return nil, err
 	}
+
+	// Ensure we close the file if we've errored
+	defer func() {
+		if retErr != nil {
+			f.Close()
+		}
+	}()
+
 	// Mark the file for removal (Linux won't remove it until we close the file)
 	if err := os.Remove(name); err != nil {
 		return nil, err
 	}
 	buf := grpcutil.GetBuffer()
 	defer grpcutil.PutBuffer(buf)
-	if _, err := io.CopyBuffer(f, objR, buf); err != nil {
+	copiedBytes, err := io.CopyBuffer(f, objR, buf)
+	if err != nil {
 		return nil, err
 	}
 	if _, err := f.Seek(0, 0); err != nil {
 		return nil, err
 	}
+
+	// Double-check that the size we got back from object storage was what we requested
+	if size != 0 && copiedBytes != int64(size) {
+		if copiedBytes > int64(size) {
+			// If the size is too large, we can truncate, but this is a worrying thing to have happen
+			logrus.Warnf("downloaded tree size is larger than requested, truncating file, size: %v, actual: %v", size, copiedBytes)
+			if err := f.Truncate(int64(size)); err != nil {
```

Comment:
Isn't this case an indication of a corrupted tree? If we truncate it, the next time around the same download will fail in different ways.

Reply:
No, this has nothing to do with the contents of the data, but if the object client returns what we requested.

Reply:
I see the size - 1 in the client code which fixes the bug we hit. What other scenario do we expect to copiedBytes > size?

Reply:
It seems like this type of validation should be a wrapper for the object storage clients, not a special case for this call site. Also, the truncation just seems like a bad idea to me. I guess you are thinking that the bytes requested could be valid, but I think it would be better to error if something like this happens.

Reply:
Yeah, I waffled about that for a bit, and I think that would be nice, but because `Reader` returns an `io.ReadCloser`, that might be a little awkward.  I guess we could wrap the reader to make sure it doesn't return too many or EOF before the requested size.